support/BRFileService.c

```diff
@@ -127,34 +228,158 @@ fileServiceCreate (const char *basePath,
+    status = sqlite3_prepare_v2 (fs->sdb, FILE_SERVICE_SDB_INSERT_ENTITY, -1, &fs->sdbInsertStmt, NULL);
+    if (SQLITE_OK != status)
+        return fileServiceCreateReturnError (fs, 1, (BRFileServiceError) {
+            FILE_SERVICE_SDB,
+            { .sdb = { status }}
+        });
+
+    // Create the SQLITE "Select Entity By Hash' Statement
+    status = sqlite3_prepare_v2 (fs->sdb, FILE_SERVICE_SDB_QUERY_ENTITY, -1, &fs->sdbSelectStmt, NULL);
+    if (SQLITE_OK != status)
+        return fileServiceCreateReturnError (fs, 1, (BRFileServiceError) {
+            FILE_SERVICE_SDB,
+            { .sdb = { status }}
+        });
+
+    // Create the SQLITE "Select Entity ' Statement
+    status = sqlite3_prepare_v2 (fs->sdb, FILE_SERVICE_SDB_QUERY_ALL_ENTITY, -1, &fs->sdbSelectAllStmt, NULL);
+    if (SQLITE_OK != status)
+        return fileServiceCreateReturnError (fs, 1, (BRFileServiceError) {
+            FILE_SERVICE_SDB,
+            { .sdb = { status }}
+        });
+
+    status = sqlite3_prepare_v2 (fs->sdb, FILE_SERVICE_SDB_UPDATE_ENTITY, -1, &fs->sdbUpdateStmt, NULL);
+    if (SQLITE_OK != status)
+        return fileServiceCreateReturnError (fs, 1, (BRFileServiceError) {
+            FILE_SERVICE_SDB,
+            { .sdb = { status }}
+        });
+
+    status = sqlite3_prepare_v2 (fs->sdb, FILE_SERVICE_SDB_DELETE_ENTITY, -1, &fs->sdbDeleteStmt, NULL);
+    if (SQLITE_OK != status)
+        return fileServiceCreateReturnError (fs, 1, (BRFileServiceError) {
+            FILE_SERVICE_SDB,
+            { .sdb = { status }}
+        });
+
+    status = sqlite3_prepare_v2 (fs->sdb, FILE_SERVICE_SDB_DELETE_ALL_TYPE_ENTITY, -1, &fs->sdbDeleteAllTypeStmt, NULL);
+    if (SQLITE_OK != status)
+        return fileServiceCreateReturnError (fs, 1, (BRFileServiceError) {
+            FILE_SERVICE_SDB,
+            { .sdb = { status }}
+        });
+
+    pthread_mutex_unlock (&fs->lock);
+
+    // Allocate the `entityTypes` array
+    array_new (fs->entityTypes, FILE_SERVICE_INITIAL_TYPE_COUNT);
+
+#if defined(DEBUG)
+    if (needSQLiteCompileOptions) {
+        needSQLiteCompileOptions = 0;
+        printf ("SQLITE ThreadSafe Mutex: %d\n", sqlite3_threadsafe());
+        printf ("SQLITE Compile Options:\n");
+        const char *option = NULL;
+        for (int index = 0;
+             NULL != (option = sqlite3_compileoption_get(index));
+             index++) {
+            printf ("-DSQLITE_%s\n", option);
+        }
+    }
+#endif
     return fs;
 }
 
+// This is callable with a partially allocated BRFileService.  So, be
+// careful with fields that might not yet exist.
 extern void
 fileServiceRelease (BRFileService fs) {
-    size_t typesCount = array_count(fs->entityTypes);
-    for (size_t index = 0; index < typesCount; index++)
-        fileServiceEntityTypeRelease (&fs->entityTypes[index]);
+    pthread_mutex_lock (&fs->lock);
+    if (NULL != fs->entityTypes) {
+        size_t typesCount = array_count(fs->entityTypes);
```

Comment:
There is the new `array_apply` and `array_free_all` that could be used here, if you were interested.

Reply:
Tried just now; compiler reported an error.

Reply:
Ah, like a mismatch if you have an array of structs (instead of pointers) and the release function expects pointers. No big deal, was worth a shot.